core/src/main/scala/com/gu/AppIdentity.scala

```diff
@@ -33,52 +33,44 @@ object AppIdentity {
       None
   }
 
-  private def fromEC2Tags(credentials: => AWSCredentialsProvider): Option[AppIdentity] = {
-
-    def listTags(instanceId: String, ec2Client: AmazonEC2): Map[String, String] = {
-      val tags = safeAwsOperation(s"Failed to describe the tags of the instance $instanceId") {
-        val result = ec2Client.describeTags(new DescribeTagsRequest().withFilters(
-          new Filter("resource-type").withValues("instance"),
-          new Filter("resource-id").withValues(instanceId)
-        ))
-        result.getTags.asScala.map{td => td.getKey -> td.getValue }.toMap
-      }
-      tags.getOrElse(Map.empty)
-    }
-
-    def ec2Client(region: Region, credentials: AWSCredentialsProvider): AmazonEC2 = {
-      val builder = AmazonEC2ClientBuilder.standard()
-      builder.setRegion(region.getName)
-      builder.setCredentials(credentials)
-      builder.build()
+  def fromASGTags(instanceId: String, credentials: => AWSCredentialsProvider): AwsIdentity = {
+    // We read tags from the AutoScalingGroup rather than the instance itself to avoid problems where the
+    // tags have not been applied to the instance before we start up (they are eventually consistent)
+    def withOneOffAsgClient[T](f: AmazonAutoScaling => T): T = {
+      val asgClient: AmazonAutoScaling = AmazonAutoScalingClientBuilder
+        .standard()
+        .withRegion(region)
+        .withCredentials(credentials)
+        .build()
+      val returned = f(asgClient)
+      asgClient.shutdown()
+      returned
     }
 
-    val region = safeAwsOperation("Failed to identify the regionName of the instance")(Regions.getCurrentRegion)
-    val tags = for {
-      awsRegion <- region
-      client <- safeAwsOperation("Failed to instantiate ec2 client")(ec2Client(awsRegion, credentials))
-      instanceId <- safeAwsOperation("Failed to identify the instanceId")(EC2MetadataUtils.getInstanceId)
-    } yield {
-      val result = listTags(instanceId, client)
-      client.shutdown()
-      result
+    def getTags(asgClient: AmazonAutoScaling): Map[String, String] = {
+      val autoscalingGroupName = {
+        val describeAutoScalingInstancesRequest = new DescribeAutoScalingInstancesRequest().withInstanceIds(instanceId)
+        val describeAutoScalingInstancesResult = asgClient.describeAutoScalingInstances(describeAutoScalingInstancesRequest)
+        describeAutoScalingInstancesResult.getAutoScalingInstances.asScala.head.getAutoScalingGroupName
+      }
+      val tags: Map[String, String] = {
```

Comment:
I don't think you need to wrap the following block in brakets and put it in a local variable as the last statement of the block is what you return. It avoids the name collision as well (though it obviously compiles fine)